Optim-wip: Add model linearization, and expanded weights spatial positions

ProGamerGov opened the pull request:
* Optionally replace nonlinear MaxPool2d layers with their linear AvgPool2d equivalents. By removing nonlinear operations from models, the expanded weights can be used for direction objectives and other stuff. I also added a few sentences to the tutorial about removing nonlinear operations.

* Optionally replace nonlinear ReLU / RedirectedReLU layers with empty layers that do nothing.

* Added info for how to visualize expanded weights spatial positions in the weight vis tutorial.

* Added info to tutorial for how to find and visualize top neuron connections by using expanded weights.

* Improved some weight vis tutorial descriptions per https://github.com/pytorch/captum/pull/543#pullrequestreview-558180872

* I also added an expanded weights test which checks that removing nonlinear layers has the correct effect on expanded weights. This also lets me assert the output tensor values.

ProGamerGov commented:
@NarineK I tried to improve the descriptions in the weight vis tutorial, so it should hopefully be easier to understand and follow now. Let me know if any areas still need improvements!

ProGamerGov commented:
@NarineK So, I was testing the crop function on expanded weights and I discovered that it has an issue with an output shape of (5,5) and an input shape of (14,14). This is a bit of a problem as the mixed4a, b, c, d, & e layers have a size of 14,14 with a size of (5,5) when the padding is cropped away. The center crop seems to work with literally every other size and output shape combo I tried, so I'm not sure what to do?


```
import torch
import torch.nn as nn
import torch.nn.functional as F

def crop_x(x, crop_list, pading_offset=True):
    h, w = x.shape
    h_crop = h - int(round((h - crop_list[0]) / 2.0))
    w_crop = w - int(round((w - crop_list[1]) / 2.0))
    return x[h_crop - crop_list[0] : h_crop, w_crop - crop_list[1] : w_crop]

px = F.pad(x, (3, 4, 3, 4), value=float("-inf"))
print(px.shape)
x_out = crop_x(px, [5, 5])
print(x_out)

px = F.pad(x, (4, 5, 4, 5), value=float("-inf"))
print(px.shape)
x_out = crop_x(px, [5, 5])
print(x_out)

px = F.pad(x, (5, 6, 5, 6), value=float("-inf"))
print(px.shape)
x_out = crop_x(px, [5, 5])
print(x_out)

px = F.pad(x, (11, 12, 11, 12), value=float("-inf"))
print(px.shape)
x_out = crop_x(px, [5, 5])
print(x_out)

px = F.pad(x, (5, 4, 4, 5), value=float("-inf"))
print(px.shape)
x_out = crop_x(px, [5, 5])
print(x_out)
```


And the output of the above code:

```
torch.Size([12, 12])
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])
torch.Size([14, 14])
tensor([[1., 1., 1., 1., -inf],
        [1., 1., 1., 1., -inf],
        [1., 1., 1., 1., -inf],
        [1., 1., 1., 1., -inf],
        [-inf, -inf, -inf, -inf, -inf]])
torch.Size([16, 16])
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])
torch.Size([28, 28])
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])
torch.Size([14, 14])
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [-inf, -inf, -inf, -inf, -inf]])
```

NarineK commented:
@ProGamerGov, that's interesting! 
We can fix the first problem if we use math.ceil instead of rand
```
    h_crop = h - int(math.ceil((h - crop_list[0]) / 2.0))
    w_crop = w - int(math.ceil((w - crop_list[1]) / 2.0))
```
Center cropping can be challenging in terms of on which side to include the endpoint. We could potentially make it as an argument to center crop function such as center towards left or right.
What do you think ?

ProGamerGov commented:
@NarineK Thanks for figuring out the fix! And yeah I definitely think that we should make it so that users can choose with side to crop to! I think that would be as simple as optionally subtracting or adding one to the existing indices? Or were you thinking of a better way of doing it?

NarineK commented:
> @NarineK Thanks for figuring out the fix! And yeah I definitely think that we should make it so that users can choose with side to crop to! I think that would be as simple as optionally subtracting or adding one to the existing indices? Or were you thinking of a better way of doing it?

yes, if the cropped sides are equal. If they are even and the cropping falls right into those equal sides then we want to make sure that we don't do unnecessary + /- 1 on one or other side.

ProGamerGov commented:
@NarineK I've added the new center crop parameter for dealing with unequal sides!

NarineK commented on captum/optim/_param/image/transform.py:
```diff
@@ -135,14 +135,22 @@ class CenterCrop(torch.nn.Module):
             values as the number of pixels from the tensor's edge, or an
             exact shape in the center.
+        offset_left (bool, optional): If the cropped away sides are not
+            equal in size, offset to the left. Default is set to False
```
nit: offset to the left means that the +1 will be on the right side ? It would be good to describe it here on an example.

ProGamerGov replied:
Okay I changed the description to:

```
offset_left (bool, optional): If the cropped away sides are not
    equal in size, offset center by +1 to the left and/or top.
    Default is set to False. This parameter is only valid when
    pixels_from_edges is False.
```

NarineK replied:
Thank you!

ProGamerGov commented:
@NarineK I added this line to the tutorial to prevent the PyTorch user warnings from appearing:

```
import warnings
warnings.filterwarnings("ignore", category=UserWarning, module='torch')
```

I think that it'll make the outputs look better if we add those lines to the tutorials, and that way we don't have to make sure to remove the warnings each time we update a notebook.

NarineK commented:
> @NarineK I added this line to the tutorial to prevent the PyTorch user warnings from appearing:
> 
> ```
> import warnings
> warnings.filterwarnings("ignore", category=UserWarning, module='torch')
> ```
> 
> Those lines should make it so that we don't have to remove the warnings every time we update a notebook.

I see - it would be good that we don't alway hide the warnings when we run the notebook in case there are important warnings that we need to be aware of. That would make sense to use before we push the code into the codebase.

ProGamerGov commented:
@NarineK Ah, okay! I've removed the import. I'll re-add and then remove it before uploading notebook updates in the future.

NarineK commented:
Thank you for the PR @ProGamerGov!
 
In terms of replacing non-linearities with the linear comparable equivalents - I think that it is not very clear why we need to do that. Can we have an explanation what is the problem with the non-linearities such as ReLUs in case we want to look at the expanded weights ? 

Also, instead of doing replacements it will be cleaner if we create an instance of the model with a flag that says replace_non_linears_with_linear_equivalents or something like that.

ProGamerGov commented:
@NarineK From the draft Lucid notebook, this is the reasoning behind removing non-linearities: https://colab.research.google.com/drive/13UZ02rmNirhQ2WI6QSFH8MphZHeyPdA_#scrollTo=-vL6Uycfjsk0&line=41&uniqifier=1

I'll have to figure out how to potentially reword it for PyTorch users to more easily understand.


For the linear model instance, do you mean adding the option in the factory function to load a linear version of the model?


I also asked Distill.pub if they want to use a variation of the weight viz notebook in an upcoming article, and they seemed interested!

NarineK commented:
> @NarineK From the draft Lucid notebook, this is the reasoning behind removing non-linearities: https://colab.research.google.com/drive/13UZ02rmNirhQ2WI6QSFH8MphZHeyPdA_#scrollTo=-vL6Uycfjsk0&line=41&uniqifier=1
> 
> I'll have to figure out how to potentially reword it for PyTorch users to more easily understand.
> 
> For the linear model instance, do you mean adding the option in the factory function to load a linear version of the model?
> 
> I also asked Distill.pub if they want to use a variation of the weight viz notebook in an upcoming article, and they seemed interested!

- Expanded weights compute the gradients of one layer w.r.t. to another layer and we technically can do that. If the first layer was input features and the second layer the last layer then it becomes feature importance (Saliency, not all interactions are captured here, just the sensitivity scores) for each output class if I understand this correctly. 
It would be good to actually have visualization for linear and non-linear and try to explain it. Right now it is unclear why we need to do that. I'll read the description in the notebook.

- w.r.t. model instance - yes, having that option in the factory function would be great!

- regarding distill.pub publication - that sounds great to me 👍

ProGamerGov commented:
@NarineK I just spoke with Chris and he explained the reasoning behind expanded weights more clearly. As you know they let us see connects between neurons that aren't literally adjacent. For Inception models, the bottleneck layers tend to produce a lot of 
superposition neurons, which other techniques are not as well equipped to handle.

So, I accidentally deleted the equation example he used when I closed the tab, but it was something like this:

```
f = relu(w * relu(w* x)* x)
f = w * w * x # Can't remember exactly if 'x' was on this line or not
```

Chris was also interested in any issues you had with understanding the weight vis article, as they are looking to explain it better.

NarineK commented:
I see - so in case of the linearity we can simply multiply the weight matrices. In case of non-linearity such as relu we need to take into consideration masked neurons by the relu and that will reflect non-linear interactions. Do we actually need to replace all non-linearities with linear equivalents or could we replace in the path from target1 to target2 because we are interested visualizing linear interactions only in the path from target1 to target2 ?

ProGamerGov commented:
@NarineK I think that could be the case, but I'm double checking with Chris!

ProGamerGov commented:
@NarineK So, Chris is pretty busy right now, but I'll let you know when he gets back to me.

In the meantime I have discovered a major bug in the optim module. The first image was made with Captum, and the second one was made with Lucent:

![captum_mixed4a_c143](https://user-images.githubusercontent.com/10626398/104136413-26022d80-5353-11eb-94d8-d503730be8a8.png)
![lucent_mixed4a_c143](https://user-images.githubusercontent.com/10626398/104136414-269ac400-5353-11eb-928b-e3be823e56bb.png)

The same model, and layer + channel (mixed4a, channel 143) were used. I've looked through pretty much every Captum function / class that could be causing the colorspace to shift to BGR, and so far I've found nothing.

ProGamerGov commented:
@NarineK 

> Do we actually need to replace all non-linearities with linear equivalents or could we replace in the path from target1 to target2 because we are interested visualizing linear interactions only in the path from target1 to target2 ?

This is Chris' reply:

> As long as you truly make everything linear, nothing that isn't between the two layers matters

ProGamerGov commented:
@NarineK Was the 'update branch' button removed?

NarineK commented:
> @NarineK Was the 'update branch' button removed?

On the github ? We haven't changed anything. Was that appearing on web page ? I usually don't do the updates on the web page but command line. Can you send a screenshot ?

ProGamerGov commented:
@NarineK Previously the web browser version of Github would display a message like this for updating PRs when their target branch got updated:

![update_branch](https://user-images.githubusercontent.com/10626398/104230174-75effb80-540a-11eb-94e1-a3e5dfeef9a1.PNG)

Not sure why it didn't show up for this PR, so I did it manually.

ProGamerGov commented:
@NarineK Captum (specifically a variation of the weight visualization tutorial notebook) is officially going to be used in the upcoming Distill.pub article!

NarineK commented:
> @NarineK Captum (specifically a variation of the weight visualization tutorial notebook) is set to be used in the upcoming Distill.pub article!

That's awesome! Do you have access to the draft of the article ?

ProGamerGov commented:
@NarineK It hasn't been added the to draft yet as they have to change the "Reproduce in a colab notebook" buttons to allow for the selection of PyTorch (Captum) or TensorFlow (Lucid). But I can let you know when the draft is updated!

NarineK commented:
> @NarineK It hasn't been added the to draft yet as they have to change the "Reproduce in a colab notebook" buttons to allow for the selection of PyTorch (Captum) or TensorFlow (Lucid). But I can let you know when the draft is updated!

That would be great! Thank you!

NarineK commented on captum/optim/_models/inception_v1.py:
```diff
@@ -50,7 +59,15 @@ def googlenet(
+            model_utils.replace_layers(model)
+        elif replace_nonlinears_with_linear_equivalents:
+            model_utils.max2avg_pool2d(model)
+            model_utils.ignore_layer(model, model_utils.ReluLayer)
```
Thank you for making the changes, @ProGamerGov! I was thinking that instead of making the replacements we can create the instance of `InceptionV1` with an argument e.g. `use_linear_modules_only`. In this case we can avoid doing replacements and calling `ignore_layer` function. Current implementation is not very different from the original implementation. The replacement piece looks like a model patch and I was thinking to construct the model with linear modules if there is such a kwargs specified. Basically, skipping relus and using AveragePool instead of maxpool when constructing the model.
Could we do that ? `https://github.com/pytorch/captum/blob/441b751150153fc423260e4419f8d7ac90626e6c/captum/optim/_models/inception_v1.py#L53`

ProGamerGov replied:
@NarineK I've made the changes!

NarineK commented on tests/optim/utils/test_circuits.py:
```diff
@@ -10,41 +10,70 @@
+                + " due to insufficient Torch version."
+            )
+
+        if torch.__version__ == "1.3.0":
```
is this `torch.__version__ == "1.3.0"` or `torch.__version__ <= "1.3.0"`  ?

ProGamerGov replied:
Oops, yeah I forgot to add the less than sign!

NarineK reviewed (commented):
Thank you for addressing the comments, @ProGamerGov !
I did a quick pass over the code and left couple nits except the tutorial. I'll do another pass on the tutorial today.

It looks like currently we don't use helper layer replacement functions. I'd suggest that we at least make them private (underscore them). Also, the user has to write their own replacement function if any customization is needed. If the functionality isn't generic enough it is better to let users to write their own replacement functions and we can keep the API more compact.

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -168,3 +172,93 @@ def collect_activations(
+
+def max2avg_pool2d(model, value: Optional[Any] = float("-inf")) -> None:
+    """
+    Replace all nonlinear MaxPool2d layers with their linear AvgPool2d equivalents.
```
> It looks like here we also do replacement of layers but don't use replace_layers function ?

Yeah I had to make a separate function because the `replace_layers` function doesn't copy the layer parameters.

NarineK replied:
`replace_layers` is a generic name if it doesn't replace any layer then we can probably give a more specific name to it  or actually try to access all parameters and copy all parameters from one module to another.

ProGamerGov replied:
Okay, I think I've figured out to replace layers while attempting to copy over any parameters that are shared!

ProGamerGov replied:
I'll make the change in a future PR, as the current one is focused on weight visualization.

ProGamerGov replied:
@NarineK Let me know what you think of the new `replace_layers` function!

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -168,3 +172,93 @@ def collect_activations(
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.avgpool(x)
+        if self.value is not None:
+            x[x == self.value] = 0.0
```
@NarineK Yeah, the primary use case is to replicate the behavior of `nn.MaxPool2d` which automatically turns the -inf values into zeros.

ProGamerGov replied:
I wasn't sure what to called the `AvgPool2dLayer` so I just left the name as it was.

NarineK replied:
Perhaps something like: `AvgPool2dConstrained`. Layer is perhaps unnecessary per PT conventions.

ProGamerGov replied:
`AvgPool2dLayer` was changed to `AvgPool2dConstrained` in one of the commits.

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -168,3 +172,93 @@ def collect_activations(
+        return x
+
+
+def max2avg_pool2d(model, value: Optional[Any] = float("-inf")) -> None:
```
Yes, they're currently only used in test cases after I changed the model definition to be build with the linear `AvgPool2dLayer` layer.

ProGamerGov replied:
The `max2avg_pool2d` function is meant for users needing to replace any nonlinear `nn.MaxPool2d` layers in their model with the right linear `nn.AvgPool2d` equivalent.

NarineK replied:
perhaps we want to have replace prefix in the function so that people understand that it is doing replacement.

ProGamerGov replied:
@NarineK Let me know what you think of the new `replace_layers` function!

ProGamerGov replied:
Also, do you have any suggests for what to change the `max2avg_pool2d` function's name to?

NarineK replied:
Perhaps something like `replace_max_with_avgconst_pool2d` or `replace_maxpool2d_with_avgconstpool2d` ?

ProGamerGov replied:
Thank you! I've changed the function name to `replace_max_with_avgconst_pool2d`!

ProGamerGov commented on captum/optim/_models/inception_v1.py:
```diff
@@ -79,9 +109,9 @@ def __init__(
-        self.localresponsenorm1 = model_utils.LocalResponseNormLayer(*lrn_vals)
+        self.conv1_relu = activ()
+        self.pool1 = pool(kernel_size=3, stride=2, padding=0)
+        self.localresponsenorm1 = LocalResponseNormLayer(*lrn_vals)
```
Okay, I've made the change!

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -168,3 +172,93 @@ def collect_activations(
+        return x
+
+
+def ignore_layer(model, layer) -> None:
```
Yeah, both `replace_layers` and `skip_layer` are meant to help users in getting their model read for Captum in some way (adding redirected relu, removing nonlinear layers, etc...).

NarineK replied:
Do we already have a `skip_layer` function somewhere else ? My suggestion was to rename `ignore_layer` to `skip_layer` because it might sounds better.

ProGamerGov replied:
@NarineK The `skip_layer` function was just the `replace_layers` with `SkipLayer` as the layer argument, so I removed it. I can re-add it if you think it's useful to keep so that users don't have to import both `replace_layers` and `SkipLayer`.

ProGamerGov replied:
Actually, I think it would be useful to keep, so I'll re-add it.

NarineK replied:
@ProGamerGov, it's good to reduce the redundancy. I thought that Skip is a better synonym of Ignore. I didn't see that you originally removed a function called `skip_layer`.

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -168,3 +172,93 @@ def collect_activations(
+        ceil_mode (bool, optional): Whether to use ceil or floor for
+            creating the output shape.
+        value (Any): Used to return any padding that's meant to be ignored
+            by pooling layers back to zero.
```
Okay, I've tried to make it more clear that the value argument refers to padding added in a previous layer.

ProGamerGov commented on tests/optim/utils/test_model_utils.py:
```diff
@@ -317,5 +317,56 @@ def test_collect_activations(self) -> None:
+
+class TestMax2AvgPool2d(BaseTest):
+    def test_max2avg_pool2d(self) -> None:
+        model = torch.nn.Sequential(
```
The function takes a model as an input, so I create the simplest model possible for testing with just a single layer.

ProGamerGov commented on tests/optim/models/test_models.py:
```diff
@@ -4,25 +4,63 @@
 from tests.helpers.basic import BaseTest, assertTensorAlmostEqual
 
 
+def check_layer_not_in_model(self, model, layer) -> None:
```
Okay, I've made the function private!

ProGamerGov commented on tests/optim/models/test_models.py:
```diff
@@ -4,25 +4,63 @@
+def check_layer_not_in_model(self, model, layer) -> None:
+    for name, child in model._modules.items():
+        if child is not None:
+            self.assertNotIsInstance(child, layer)
```
It doesn't look as just adding an bool parameter to replace `assertNotIsInstance` with `assertIsInstance` if we want to check that a layer exists in the model. So, I created a second function that does that!

ProGamerGov commented on captum/optim/_models/inception_v1.py:
```diff
@@ -50,27 +69,38 @@ def googlenet(
         else:
             state_dict = torch.load(model_path, map_location="cpu")
         model.load_state_dict(state_dict)
-        model_utils.replace_layers(model)
```
The `replace_layers` function is not being used externally in any tutorials right now. And yes it was replaced in the model factory function by setting the activations internally like the pooling layers.

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -168,3 +172,93 @@ def collect_activations(
+            max2avg_pool2d(child)
+
+
+class IgnoreLayer(torch.nn.Module):
```
Yeah, I like that name better! I've made the change!

ProGamerGov commented:
@NarineK The `skip_layer` function is redundant as it just calls replace_layers with `SkipLayer` as the layer argument, so I'll remove it. But for  `max2avg_pool2d` & `replace_layers` I feel like they are both useful & general enough to keep. They also show the user how they can create their own custom layer replacement functions. Though we could move them to some sort of other utils submodule.

NarineK commented on captum/optim/_utils/models.py:
```diff
@@ -168,3 +223,89 @@ def collect_activations(
+        return x
+
+
+def skip_layers(model, layers) -> None:
```
nit: typehint

ProGamerGov replied:
So, I'm struggling to figure out the type hints to use for models, layer instances, and layer types. I found a StackOverflow post mentioning that `typing.Type`, but it doesn't work very well. So in the meantime I've set the type hints as `Any`.

ProGamerGov replied:
@NarineK Elsewhere on Captum model variables aren't type hinted for a similar reason I think.

NarineK replied:
`model` and `layers` are nn.Module type 
https://github.com/pytorch/captum/blob/master/captum/attr/_core/deep_lift.py#L108

ProGamerGov replied:
@NarineK Thank you! I've added all the missing type hints!

NarineK commented on captum/optim/_utils/models.py:
```diff
@@ -70,22 +71,76 @@ class ReluLayer(nn.Module):
+        for i in range(len(l2_vars))
+        if str(l2_vars[i]) != "self"
+    ]
+    l2_vars = [p.split(":")[0] if ":" in p and "=" not in p else p for p in l2_vars]
```
I was just trying this simple case:
```
class C:
    def myfunc(self, a, b:int=10, c=20, d=30):
        pass
l2_vars = list(signature(C().myfunc.__init__).parameters.values())
.....
>> [<Parameter "*args">, <Parameter "**kwargs">]
```
When do you see the output with ':' and '=' symbols ?

ProGamerGov replied:
@NarineK I saw the symbols when replacing MaxPool2d with AvgPool2d layers.

NarineK replied:
Ah, I think in my case it should be:
```
l2_vars = list(signature(C().myfunc).parameters.values())
```
I was calling  __init__.

ProGamerGov replied:
Ah, okay!

NarineK replied:
For this example it looks like we are losing value: 10?
```
class C:
    def myfunc(self, a, b:int=10, c=20, d=30):
        pass

l2_vars = list(signature(C().myfunc).parameters.values())
l2_vars = [
        str(l2_vars[i]).split()[0]
        for i in range(len(l2_vars))
        if str(l2_vars[i]) != "self"
    ]
l2_vars = [p.split(":")[0] if ":" in p and "=" not in p else p for p in l2_vars]

```
.....
['a', 'b', 'c=20', 'd=30']

ProGamerGov replied:
@NarineK Default values that don't overlap between layers can be set via kwargs, or they remain what they are set as when using the `_transfer_layer_vars` function. If the layer we are transfering values from has a `b`, then the value is overridden. If `b` doesn't exist, it will set to 10 in the layer instance.

NarineK replied:
Ah I see and we assume that `shared_vars` and `kwargs` will have matching sizes

NarineK commented on tests/optim/models/test_models.py:
```diff
@@ -1,28 +1,77 @@
+    def check_for_layer_in_model(model, layer, in_model: List) -> None:
+        for name, child in model._modules.items():
+            if child is not None:
+                if isinstance(child, layer):
```
Could we have something like this instead?

```
    def check_for_layer_in_model(model, layer, in_model: List) -> None:
        for name, child in model._modules.items():
            if child is not None:
                if isinstance(child, layer):
                    return
                check_for_layer_in_model(child, layer, in_model)

    check_for_layer_in_model(model, layer, in_model)
    # If we reach here then there was no True 
    self.assertTrue(False)
```

ProGamerGov replied:
@NarineK Is your code meant to return True after finding the first layer, or else it returns False?

NarineK replied:
yes, it basically passes the assert and returns since the function doesn't return anything I didn't return anything otherwise if we reach `self.assertTrue(False)` it should always fail, basically your original any has all false values

ProGamerGov replied:
Okay, I've replaced the function with your improved version!

ProGamerGov replied:
So your version had an error, and I realized my version could also optional check for a specified number of layers. I reverted back to my function and added the addition parameter. The new parameter is not used at the moment, but in the future it can be used to check that all layers that should have been replaced, were replaced.

NarineK replied:
I see - what was the error ?

ProGamerGov replied:
@NarineK The inner function only returns to the outer function, and not the test function. So, it always fails.

NarineK replied:
Ah, sorry, I missed the outer function.
Something like this might suffice too. 
```
def _check_layer_in_model(self, model, layer) -> None:
    def check_for_layer_in_model(model, layer, in_model: List) -> bool:
        for name, child in model._modules.items():
            if child is not None:
                if isinstance(child, layer):
                    return True
                check_for_layer_in_model(child, layer, in_model)
        return False

    self.assertTrue(check_for_layer_in_model(model, layer, in_model))

```

ProGamerGov replied:
@NarineK Thank you!

NarineK replied:
@ProGamerGov, you might need the return there:
https://github.com/pytorch/captum/blob/ba076856d14a4c85f5903e371b272ac20b293c07/tests/optim/models/test_models.py#L27
I didn't test locally

ProGamerGov replied:
@NarineK I tested locally before adding it!

NarineK replied:
I think in this case it works because it doesn't go to next recursive call.
I've created this simple example:
```
def a():
    i = 0
    def b():
         nonlocal i
         i += 1 
         if i < 6:
            if i == 4:
                return True
            # doesn't return b 
            b()
         return False

    print(b())    

a()
# Always return False
```

NarineK reviewed (commented):
Thank you, @ProGamerGov ! I added couple more nits. Replied to your comments.

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -70,22 +71,76 @@ class ReluLayer(nn.Module):
+
+def _transfer_layer_vars(layer1, layer2, **kwargs):
+    """
+    Given a layer instance, create a new layer instance of another class
```
Okay, I've changed the description to refer to both layers!

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -70,22 +71,76 @@ class ReluLayer(nn.Module):
+            initialization variables from.
+        layer2: (nn.Module): The layer class to create with the variables
+            from of layer1.
+        transfer_vars (bool, optional): Wether or not to try and copy
```
Fixed!

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -70,22 +71,76 @@ class ReluLayer(nn.Module):
+        layer2: (nn.Module): The layer class to create with the variables
+            from of layer1.
+        kwargs: (Any, optional): Any additional variables to use when creating
+            the new layer.
```
I changed it, but I now think that the original `old_layer` and `new_layer` variables names were better. So, I changed them back.

NarineK replied:
I think that `old_layer` and `new_layer` are better names too.

ProGamerGov commented:
@NarineK The multi-gpu test seems to be broken currently as @greentfrapp had the same errors.

ProGamerGov commented:
@NarineK Are you able to take a look at the notebook tutorial in the next day or two? It's important for the upcoming article.

ProGamerGov commented:
@NarineK I sent you an email regarding this PR, so please don't reply to my last comment. The email doesn't contain any info that people might need discussion wise for the Captum project, but it is important!

NarineK reviewed (approved):
Thank you for working on this PR, @ProGamerGov! Looks great! Here are couple suggestions.

1. It would be good that we add a bit more description in the beginning of the tutorial. The first 1-2 sentences from `Extracting expanded weights` section could be a good fit there too. In addition to that describing a bit about NMF etc...
2. `visualize_activations` and `vis_multi share` share code that I think could be combined together. Also, `vis_multi` is not a very clear name, maybe we can give a more clear name to it. Some of the args have type hint some not. It would be good to be consistent and have it everywhere or remove it everywhere.
3. `get_expanded_weights ` sounds a bit like a getter function but that function does more that a get. We could probably call it `extract_expanded_weights` or another name of your choice.
4. Here you might also want to describe in the text what y and x are:[target2 output channels, target1 output channels, y, x] . You can also describe that circuits are the grads of the target2 for the center activation w.r.t. target 1 etc ... 
5.  `vis_neuron_direction` seems to be a copy of `vis_multi` for a different loss where channel is replaced with vec. We can combine these functions in one `visualize_activations` function or at least keep transformation as a sharable definition.
6. For the section, `Visualizing the spatial positions of expanded weights image`, I think you can probably visualize the show(W_3a_3b_hm) next to it for easy comparison.  Are we getting identical results what OpenAI has in their implementation ?
7. Do you have documentation or citation on `Weight banding` ?
8.  In terms of Weight banding I see that you access the conv layer but not pooling. Don't we need to call circuits for W_5b_c5x5 and pooling layers ?
Pooling layer related circuits you are getting in the below cell: W_p2_3a = optimviz ... 
but that's not used in the cell above ... 
I might be missing something here
9. I think that we can abstract out this piece in a function and reuse it: A = []
       for i in highlow_units:
           x_out = vis_multi(model, model.conv3, i)
           A.append(x_out.detach())

      grid_img = torchvision.utils.make_grid(torch.cat(A), nrow=5)
     show(grid_img)
10. nit: Can we, please, also describe where are we getting the `highlow_units` and `bw` indices from ?
11. `reducer.components` are we setting the components in ChannelReducer ? https://github.com/pytorch/captum/blob/ba076856d14a4c85f5903e371b272ac20b293c07/captum/optim/_utils/reducer.py#L17
12. In the section `Multiple related neurons with a small number of factors` I'd add more docuementation describing the components and the semantics. Also couple more sentences about the highlights of the visualizations.

ProGamerGov commented:
@NarineK 

1. & 12. Currently working on improving the descriptions based on your feedback!

2. `visualize_activations` only shares two lines of code plus two transforms with `vis_multi` (now known as `vis_neuron`), so it could be better to just leave them as separate. I've also added type hints!

3. The Lucid version of `get_expanded_weights` is also called `get_expanded_weights`, but I change the name if you think that a different name would be better.

4. I've replace `y` & `x` with `height` & `width`. Also, can you elaborate more on that explanation of circuits?

5. I've merged `vis_neuron_direction` and `vis_multi` into a new function called `vis_neuron`.

6. I've added `show(W_3a_3b_hm)` to the spatial visualization cell.

7. I've added info about the upcoming weight banding paper. I've added some words that clarify this to the tutorial.

8. For weight banding we are looking directly at the non expanded weights.

9. Okay, I've turned it into a function that's reused across the tutorial!

10. I've added links to where the neurons come from.

11. `reducer.components` are a factorization matrix created during decomposition when the channel reducer is run: https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html

ProGamerGov commented:
@NarineK You can merge this PR whenever you want now!

NarineK commented:
> @NarineK
> 
> 1. Currently working on improving the descriptions based on your feedback, though it might be better to do it in a future PR as I can't think of how to word it at the moment.
> 2. `visualize_activations` (now called `vis_neuron_large`) only shares two lines of code plus two transforms with `vis_multi` (now known as `vis_neuron`), so it could be better to just leave them as separate. I've also added type hints!
> 3. The Lucid version of `get_expanded_weights` is also called `get_expanded_weights`, but I can change the name if you think that a different name would be better.
> 4. I've replaced `y` & `x` with `height` & `width`. Also, can you elaborate more on that explanation of circuits?
> 5. I've merged `vis_neuron_direction` and `vis_multi` into a new function called `vis_neuron`.
> 6. I've added `show(W_3a_3b_hm)` to the spatial visualization cell. And the weights should be the same as the Lucid / OpenAI version.
> 7. I've added info about the upcoming weight banding paper.
> 8. For weight banding we are looking directly at the non expanded weights. I've added some words that clarify this to the tutorial.
> 9. Okay, I've turned it into a function that's reused across the tutorial!
> 10. I've added links to where the neurons come from.
> 11. The `reducer.components` variable is a factorization matrix created during decomposition when the channel reducer is run: https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html
> 12. I've improved the documentation, so hopefully everything makes more sense now!

Thank you for addressing all comments @ProGamerGov!
 1. Regarding 1: Sure, we can do it in a separate PR.  I was thinking that you can summarize what you are doing in the notebook in the top of the notebook. Something like as abstract. 
3. I understand that you wanted to be consistent with lucid but I think that from engineering perspective `extract_expanded_weights` or `compute_expanded_weights` is more clear unless you want to have it identical with lucid then I'm fine with it too.

  - nit: Did we wanted to add return here:    https://github.com/pytorch/captum/blob/b9cbf02df212d64831eae552cfd0224a8d0bb1a3/tests/optim/models/test_models.py#L27  - If you want to can add more explanation to the printout of top neurons and add something like: ``` print('Top {} neurons of channel 379
 with the largest L2-norm: {} '.format(top_connected_neurons.size(), list(top_connected_neurons))) ....  ```

ProGamerGov commented:
@NarineK 

1. I've added a basic summary what we are doing in the notebook!

2. Okay, I've changed `get_expanded_weights` to `extract_expanded_weights`!

3. I've added the improved output printing for the top channels!

4. I'm not quite sure, but the current version of the test seems to work?

NarineK commented:
> @NarineK
> 
> 1. I've added a basic summary what we are doing in the notebook!
> 2. Okay, I've changed `get_expanded_weights` to `extract_expanded_weights`!
> 3. I've added the improved output printing for the top channels!
> 4. I'm not quite sure, but the current version of the test seems to work? We can change it in a future PR if need be.

Thank you for addressing the comments, @ProGamerGov ! Regarding 4. did you try to add return statement and it failed ? I think that currently there is a bug in the recursion and it might be working that now but if we want to use the function in a more general context we will see issues.

ProGamerGov commented:
@NarineK I'm not quite sure what you mean? There are the two return statements, and stops when it finds the target instance type:

```
def _check_layer_in_model(
    self,
    model: torch.nn.Module,
    layer: Type[torch.nn.Module],
) -> None:
    def check_for_layer_in_model(model, layer) -> bool:
        for name, child in model._modules.items():
            if child is not None:
                if isinstance(child, layer):
                    return True
                check_for_layer_in_model(child, layer)
        return False

    self.assertTrue(check_for_layer_in_model(model, layer))
```

NarineK commented:
> @NarineK I'm not quite sure what you mean? There are the two return statements, and stops when it finds the target instance type:
> 
> ```
> def _check_layer_in_model(
>     self,
>     model: torch.nn.Module,
>     layer: Type[torch.nn.Module],
> ) -> None:
>     def check_for_layer_in_model(model, layer) -> bool:
>         for name, child in model._modules.items():
>             if child is not None:
>                 if isinstance(child, layer):
>                     return True
>                 check_for_layer_in_model(child, layer)
>         return False
> 
>     self.assertTrue(check_for_layer_in_model(model, layer))
> ```

@ProGamerGov, I meant a return here: return check_for_layer_in_model(child, layer, in_model)

ProGamerGov commented:
@NarineK Ah, thank you! I made the change and it causes the test to fail.

ProGamerGov commented:
I added the GPU test fix as well. The CircleCI tests aren't running yet because of the concurrency limit, but it should work.

ProGamerGov commented:
@NarineK I added the GPU test fix, and now the GPU test is working again!

ProGamerGov commented:
@NarineK I tested the `_check_layer_in_model` function and it does go through every layer in the model in it's current form. So, I don't see where the bug / issue is?

NarineK commented:
> @NarineK I tested the `_check_layer_in_model` function and it does go through every layer in the model in it's current form. So, it looks like it's able to reach the inner modules.

Thank you for testing it @ProGamerGov !
It fails for example in this use case:
```
        class SubModule(nn.Module):
            def __init__(self):
                super().__init__()
                self.r = nn.ReLU()

            def forward(self, x):
                return self.r(x)

        class MyModel(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.l1 = nn.Linear(3, 4)
                self.sub = SubModule()
            def forward(self, x):
                return self.l1(self.sub(x))
```
Here: ` _check_layer_in_model(self, MyModel(), nn.ReLU)` fails.
